File: 3710. Maximum Partition Factor.java
class Solution {
    public boolean bfs(HashMap<Integer, ArrayList<int[]>> adj, int vertex, int[] color, int target) {

        Queue<Integer> q = new LinkedList<>();
        q.offer(vertex);
        color[vertex] = 0;

        while (!q.isEmpty()) {
            int u = q.poll();

            for (int[] neig : adj.get(u)) {
                int v = neig[0];
                int dist = neig[1];

                if (dist >= target) continue;

                if (color[v] == -1) {
                    color[v] = 1 - color[u];
                    q.offer(v);
                } else if (color[v] == color[u]) {
                    return false;
                }
            }
        }

        return true;
    }
    public boolean isPossible(int[][] points, int target, HashMap<Integer, ArrayList<int[]>> adj){
        int[] color = new int[points.length];
        Arrays.fill(color, -1);
        
        for(int i = 0; i < points.length; i++){
            if(color[i] == -1 && !bfs(adj, i, color, target)) return false;
        }

        return true;
    }
    public int maxPartitionFactor(int[][] points) {
        if (points.length == 2) return 0;
        
        int left = 0, right = 0;
        for (int i = 0; i < points.length; i++) {
            for (int j = i + 1; j < points.length; j++) {
                int dist = Math.abs(points[i][0] - points[j][0])
                         + Math.abs(points[i][1] - points[j][1]);

                right = Math.max(right, dist);
            }
        }

        HashMap<Integer, ArrayList<int[]>> adj = new HashMap<>();
        for(int i = 0; i < points.length; i++){
            int[] x = points[i];
            ArrayList<int[]> arr = adj.getOrDefault(i, new ArrayList<>());

            for(int j = 0; j < points.length; j++){
                if(i == j) continue;
                int[] y = points[j];

                int dist = Math.abs(x[0] - y[0]) + Math.abs(x[1] - y[1]);
                arr.add(new int[]{j, dist});
            }
            adj.put(i, arr);
        }

        int ans = 0;
        while(left <= right){
            int mid = left + (right - left) / 2;
            if(isPossible(points, mid, adj)){
                ans = mid;
                left = mid + 1;
            }
            else right = mid - 1;
        }

        return ans;
    }
}

File: 3710. Maximum Partition Factor.cpp
class Solution {
public:
    bool bfs(unordered_map<int, vector<vector<int>>>& adj, int vertex, vector<int>& color, int target) {
        queue<int> q;
        q.push(vertex);
        color[vertex] = 0;

        while (!q.empty()) {
            int u = q.front();
            q.pop();

            for (vector<int>& neig : adj[u]) {
                int v = neig[0];
                int dist = neig[1];

                if (dist >= target) continue;

                if (color[v] == -1) {
                    color[v] = 1 - color[u];
                    q.push(v);
                }
                else if (color[v] == color[u]) {
                    return false;
                }
            }
        }

        return true;
    }

    bool isPossible(vector<vector<int>>& points, int target, unordered_map<int, vector<vector<int>>>& adj) {
        vector<int> color(points.size(), -1);

        for (int i = 0; i < points.size(); i++) {
            if (color[i] == -1 && !bfs(adj, i, color, target))
                return false;
        }

        return true;
    }

    int maxPartitionFactor(vector<vector<int>>& points) {
        if (points.size() == 2) return 0;

        int left = 0, right = 0;

        for (int i = 0; i < points.size(); i++) {
            for (int j = i + 1; j < points.size(); j++) {
                int dist = abs(points[i][0] - points[j][0])
                         + abs(points[i][1] - points[j][1]);

                right = max(right, dist);
            }
        }

        unordered_map<int, vector<vector<int>>> adj;

        for (int i = 0; i < points.size(); i++) {
            vector<vector<int>> arr;

            for (int j = 0; j < points.size(); j++) {
                if (i == j) continue;

                int dist = abs(points[i][0] - points[j][0])
                         + abs(points[i][1] - points[j][1]);

                arr.push_back({j, dist});
            }

            adj[i] = arr;
        }

        int ans = 0;

        while (left <= right) {
            int mid = left + (right - left) / 2;

            if (isPossible(points, mid, adj)) {
                ans = mid;
                left = mid + 1;
            }
            else {
                right = mid - 1;
            }
        }

        return ans;
    }
}; 